### DISCUSSION TITLE
SparseMatrix is equal to the dense one

### DISCUSSION THREAD

**Discussion by AlexandreAlvesDB** [josdejong/mathjs#2830] [Q&A]
Hello everyone, 

i'm trying to implement a solver for the least square problem and I'm running into an issue.
I want a compressed version of a matrix which is full of zeros so I used mathjs.sparse 

![image](https://user-images.githubusercontent.com/91118126/198986512-12b5f584-e059-433f-8dfd-76893a26a068.png)

But what is returned is a square matrix with zero values, maybe I understood it wrong.

![image](https://user-images.githubusercontent.com/91118126/198986773-428ed5a0-3294-4e87-b513-12e0143898d6.png)

If someone could help me on this one, I'd be grateful !

**Comment by gwhitney** [comment]
Hmm, this is odd. A SparseMatrix should only store entries `m` where `equalScalar(m, 0)` is false. Can you test some of the entries you believe are zero with this test before you form the SparseMatrix? If they return false, that may mean you need to round almost-zero entries to actually zero before forming the SparseMatrix. If they return true, then it is possible you have uncovered a bug in the function that creates a SparseMatrix from  an ordinary matrix -- not certain as I am not an expert on that code. Anyhow, let us know about this test.

**Reply by josdejong** [reply]
I just tried a little example which works as expected:

```js
const d = math.matrix([[0,0,0],[0,4,0],[0,0,0]])
const s = math.sparse(d)
console.log(s._values) // [ 4 ] 
```

@AlexandreAlvesDB can you share a minimal code example demonstrating the issue so we can do some debugging?

**Reply by AlexandreAlvesDB** [reply]
I'm sorry I changed a lot the code since and I'm trying to replicate the array I had at the time to save it.

**Reply by AlexandreAlvesDB** [reply]
So I made a fiddle that demonstrates the issue :
https://jsfiddle.net/9e0dw8ko/3/

When I create the initial matrix with math.zeros, everything works fine
When the values are set to 0 using .set(), the sparse function doesn't work later

```
let matrix = math.matrix("sparse");
matrix.resize([50,30]);
 for(let i = 0; i<matrix.size()[0] ;i++)
  {
    for(let j = 0; j<matrix.size()[1] ;j++)
    {
      matrix.set([i,j],0);
      
    }
  }
  

let product = math.multiply(math.transpose(matrix), matrix);

console.log(math.sparse(product));

let matrix2 = math.matrix("sparse");
matrix2 = math.zeros([50,30]);
let product2 = math.multiply(math.transpose(matrix2), matrix2);

console.log(math.sparse(product2));
```

**Reply by gwhitney** [reply]
The problem is not the call to math.sparse() in the first console.log() line. At that point, `product` is already a sparse matrix, so math.sparse rightly does nothing. The difficulty is that `product` is already a sparse matrix of all zeros, all of which are explicitly recorded as values, i.e. not compressed away. So the problem is in one of matrix.set, math.transpose, or math.multiply. I will investigate.

**Reply by gwhitney** [reply]
The problem may be in the set method of SparseMatrix. See SparseMatrix.js in `/src/type/matrix`. The guts are:
```
    // find value index
    const k = _getValueIndex(i, this._ptr[j], this._ptr[j + 1], this._index)
    // check k is prior to next column k and it is in the correct row
    if (k < this._ptr[j + 1] && this._index[k] === i) {
      // check value != 0
      if (!eq(v, zero)) {
        // update value
        this._values[k] = v
      } else {
        // remove value from matrix
        _remove(k, j, this._values, this._index, this._ptr)
      }
    } else {
      // insert value @ (i, j)
      _insert(k, i, j, v, this._values, this._index, this._ptr)
    }
```
You see in the top-level "else" block there is never a check for zero, so it seems to me that `.set()` may be calling `_insert` with zero values, "polluting" the compression of the SparseMatrix. If that's correct, a fix might be to simply change the `else` to `else if (!eq(v, zero))` since there is no point inserting a zero into a SparseMatrix. Sorry I don't at the moment have time to try this, but if you do and it works, a PR with the fix and a test for this problem (maybe just like your fiddle but with smaller matrices) would be welcome.

**Reply by AlexandreAlvesDB** [reply]
I can confirm it was that, I wrote the unit test to check first if the build would fail (it did) then I juste added a check of the value for the _insert and well it passed.
I'm having issue with prettier to finish the test and build but as soon as It works, I will propose the fix

**Comment by AlexandreAlvesDB** [comment]
I see, I will check if that's right and do that !

Le mar. 1 nov. 2022 à 15:30, Glen Whitney ***@***.***> a
écrit :

**Comment by josdejong** [comment]
This issue has been fixed now in v11.3.3. Thanks @AlexandreAlvesDB .

**Reply by AlexandreAlvesDB** [reply]
You're welcome !
